USAM - Not finding references from Unity internals with asmdef in docker

When running services that references asmdefs, we are copying now all the scripts that it tries to compile, however in the csproj that we generate for those asmdefs, there are some references that are creating warnings when building it in docker. These are the warnings in question:
```
/usr/share/dotnet/sdk/6.0.424/Microsoft.Common.CurrentVersion.targets(2302,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "ICSharpCode.SharpZipLib". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation 
errors. [/subsrc/Library/BeamableEditor/GeneratedProjects/TestingAssembly/UNITY_GENERATED_PROJECT_TestingAssembly.csproj]
/usr/share/dotnet/sdk/6.0.424/Microsoft.Common.CurrentVersion.targets(2302,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "Unity.Analytics.Tracker". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation 
errors. [/subsrc/Library/BeamableEditor/GeneratedProjects/TestingAssembly/UNITY_GENERATED_PROJECT_TestingAssembly.csproj]
/usr/share/dotnet/sdk/6.0.424/Microsoft.Common.CurrentVersion.targets(2302,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "Unity.Analytics.Editor". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation 
errors. [/subsrc/Library/BeamableEditor/GeneratedProjects/TestingAssembly/UNITY_GENERATED_PROJECT_TestingAssembly.csproj]
/usr/share/dotnet/sdk/6.0.424/Microsoft.Common.CurrentVersion.targets(2302,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "Unity.Analytics.StandardEvents". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get 
compilation errors. [/subsrc/Library/BeamableEditor/GeneratedProjects/TestingAssembly/UNITY_GENERATED_PROJECT_TestingAssembly.csproj]

```

Right now these are just warnings, but is it possible that if someone tries to use something from thoses references it will break things in docker?

--- a/cli/cli/App.cs
+++ b/cli/cli/App.cs
@@ -364,6 +364,7 @@ public class App
 		Commands.AddSubCommand<ServicesGenerateLocalManifestCommand, ServicesGenerateLocalManifestCommandArgs, ServicesCommand>();
 		Commands.AddSubCommand<ServicesGenerateTarballCommand, ServicesGenerateTarballCommandArgs, ServicesCommand>();
 		Commands.AddSubCommand<ServicesBuildCommand, ServicesBuildCommandArgs, ServicesCommand>();
+		Commands.AddSubCommand<ServicesUpdateDockerfileCommand, ServicesUpdateDockerfileCommandArgs, ServicesCommand>();
 		// content commands
 		
 		Commands.AddRootCommand<ContentCommand>();

--- a/cli/cli/Commands/Services/ServicesUpdateDockerfileCommand.cs
+++ b/cli/cli/Commands/Services/ServicesUpdateDockerfileCommand.cs
@@ -11,6 +11,8 @@ public class ServicesUpdateDockerfileCommandArgs : CommandArgs
 
 public class ServicesUpdateDockerfileCommand : AppCommand<ServicesUpdateDockerfileCommandArgs>, IEmptyResult
 {
+	public override bool IsForInternalUse => true;
+	
 	public ServicesUpdateDockerfileCommand() : base("update-dockerfile", "Updates the Dockerfile for the specified service")
 	{
 	}

--- a/cli/cli/Commands/Services/ServicesUpdateDockerfileCommand.cs
+++ b/cli/cli/Commands/Services/ServicesUpdateDockerfileCommand.cs
@@ -1,0 +1,41 @@
+using cli.Services;
+using Serilog;
+using System.CommandLine;
+
+namespace cli;
+
+public class ServicesUpdateDockerfileCommandArgs : CommandArgs
+{
+	public string ServiceName;
+}
+
+public class ServicesUpdateDockerfileCommand : AppCommand<ServicesUpdateDockerfileCommandArgs>, IEmptyResult
+{
+	public ServicesUpdateDockerfileCommand() : base("update-dockerfile", "Updates the Dockerfile for the specified service")
+	{
+	}
+
+	public override void Configure()
+	{
+		AddArgument(new Argument<string>("ServiceName", description: "The name of the microservice to udpate the Dockerfile"),
+			(args, i) => args.ServiceName = i);
+	}
+
+	public override async Task Handle(ServicesUpdateDockerfileCommandArgs args)
+	{
+		if (!args.BeamoLocalSystem.BeamoManifest.TryGetDefinition(args.ServiceName, out var serviceDefinition))
+		{
+			Log.Error($"The service {args.ServiceName} does not have a definition in the manifest");
+			return;
+		}
+
+		if (serviceDefinition.Protocol != BeamoProtocolType.HttpMicroservice)
+		{
+			Log.Error($"The service {args.ServiceName} is not a HttpMicroservice");
+			return;
+		}
+
+		await args.BeamoLocalSystem.UpdateDockerFile(serviceDefinition);
+		Log.Information($"The service {args.ServiceName} Dockerfile has ben updated!");
+	}
+}

--- a/cli/cli/Services/BeamoLocalSystem.cs
+++ b/cli/cli/Services/BeamoLocalSystem.cs
@@ -380,6 +380,14 @@ COPY {serviceFileTag} /subsrc/{servicePathTag}";
 			newText = newText.Replace(endTag, replacementFile.Replace(servicePathTag, path).Replace(serviceFileTag, file).Replace('\\', '/').Insert(0, Environment.NewLine));
 		}
 
+		//Add all dll files in the project
+		var allDlls = ProjectContextUtil.GetAllIncludedDlls(definition, _configService);
+		foreach (var dll in allDlls)
+		{
+			var path = Path.GetDirectoryName(dll);
+			newText = newText.Replace(endTag, replacementFile.Replace(servicePathTag, path).Replace(serviceFileTag, dll).Replace('\\', '/').Insert(0, Environment.NewLine));
+		}
+
 		//Copy the services files
 		Log.Verbose($"adding service files projPath=[{projectDir}]");
 

--- a/cli/cli/Services/ProjectContextUtil.cs
+++ b/cli/cli/Services/ProjectContextUtil.cs
@@ -609,6 +609,52 @@ public static class ProjectContextUtil
 		return results;
 	}
 
+	public static List<string> GetAllIncludedDlls(BeamoServiceDefinition definition, ConfigService configService)
+	{
+		var buildEngine = new ProjectCollection();
+		Project buildProject = buildEngine.LoadProject(definition.ProjectPath);
+
+		return GetAllIncludedDlls(buildProject, configService);
+	}
+
+	public static List<string> GetAllIncludedDlls(Project project, ConfigService configService)
+	{
+		var buildEngine = new ProjectCollection();
+		var results = new List<string>();
+		const string REFERENCE_TYPE = "Reference";
+		const string PROJECT_REFERENCE_TYPE = "ProjectReference";
+
+		//First check if there are other projects being referenced here, so we run through them too
+		var projs = project.GetItemsIgnoringCondition(PROJECT_REFERENCE_TYPE).ToArray();
+		foreach (ProjectItem item in projs)
+		{
+			var name = Path.GetFileName(item.EvaluatedInclude);
+			var relativePath =
+				configService.GetPathFromRelativeToService(item.EvaluatedInclude, project.DirectoryPath);
+			var combinedPath = Path.Combine(relativePath, name);
+			var refProject = buildEngine.LoadProject(combinedPath);
+
+			var partialResult = GetAllIncludedDlls(refProject, configService);
+			results.AddRange(partialResult);
+		}
+
+		var references = project.GetItemsIgnoringCondition(REFERENCE_TYPE).ToArray();
+
+		foreach (var dllReference in references)
+		{
+			ProjectMetadata hintPathMetadata = dllReference.Metadata.FirstOrDefault(m => m.Name.Equals("HintPath"));
+			if (hintPathMetadata == null) continue;
+			var dllPath = hintPathMetadata.EvaluatedValue;
+			var name = Path.GetFileName(dllPath);
+			var relativePath =
+				configService.GetPathFromRelativeToService(dllPath, project.DirectoryPath);
+			var fixedPath = Path.Combine(relativePath, name);
+			results.Add(fixedPath);
+		}
+
+		return results;
+	}
+
 
 	public static void UpdateUnityProjectReferences(BeamoServiceDefinition definition, List<string> projectsPaths, List<string> assemblyNames)
 	{

--- a/client/Packages/com.beamable.server/Editor/Usam/Editor/CsharpProjectUtil.cs
+++ b/client/Packages/com.beamable.server/Editor/Usam/Editor/CsharpProjectUtil.cs
@@ -92,7 +92,6 @@ namespace Beamable.Server.Editor.Usam
 					var oldContent = File.ReadAllText(fileName);
 					if (oldContent.Equals(content))
 					{
-						Debug.Log($"Skipping assembly {assembly.name} because it does not have any changes");
 						continue;
 					}
 				}
@@ -158,7 +157,10 @@ namespace Beamable.Server.Editor.Usam
 			foreach (var dll in assembly.compiledAssemblyReferences)
 			{
 				if (!dll.StartsWith(projectRoot)) continue;
-				yield return dll.Substring(projectRoot.Length + 1);
+				var dllPath = dll.Substring(projectRoot.Length + 1);
+				var dllName = Path.GetFileName(dllPath);
+				if (!IsValidReference(dllName)) continue;
+				yield return dllPath;
 			}
 
 			yield break;

